asciidoc/courses/llm-fundamentals/modules/2-vectors-semantic-search/lessons/4-improving-semantic-search/lesson.adoc

```diff
@@ -0,0 +1,83 @@
+
+On the other hand, if you are recommending a childrens movie to pass away a rainy afternoon, *Jack and the Beanstalk* is worlds away from *The Shining*.
+
+
+== Incorporating Graph Features
+
+Graph features can enrich vector-backed semantic search by providing structural and relational context to data.
+Leveraging the interconnectedness inherent in graphs, semantic search can draw upon relationships and hierarchies between entities, enhancing the depth and relevance of search results.
+While vectors capture semantic nuances, the addition of graph structures allows for a more holistic understanding of data, considering both its inherent meaning and its position in a broader network of knowledge.
+
+This topic is out of the scope of this module, but for more information, we recommend watching link:https://www.youtube.com/watch?v=bRD09ndyJNs[Going Meta - Ep 21: Vector-based Semantic Search and Graph-based Semantic Search^], in which Dr Jesus Barassa and Alexander Erdl explore the differences between Vector-based Semantic Search and Graph-based Semantic Search.
+
+== Using Feedback
+
+A simpler approach may be to use User feedback to gradually re-rank and curate the content returned by a vector index search.
+This is the approach we take with the GraphAcademy chatbot.
+
+To celebrate the launch of the Vector Index, we published an article detailing link:https://medium.com/neo4j/building-an-educational-chatbot-for-graphacademy-with-neo4j-f707c4ce311b[how we used vector index to identify Neo4j content] to improve LLM responses.
+The article demonstrates how we ingested Neo4j Documentation and GraphAcademy lessons, using embeddings of the content to populate a vector index.
+
+// TODO: Move to CDN?
+image::https://miro.medium.com/v2/resize:fit:4800/format:webp/1*BpDS376rEKb6kWBlxNobRQ.png[The GraphAcademy Chatbot Data Model]
+
+When a user asks a question, the server generates an embedding of the user's question and uses the vector index to identify relevant content.
+
+This approach works well in most cases, but now and again, the index fails to return relevant content, in which case the LLM is instructed to respond to the user with an apology.
+
+When this happens, or the user provides feedback, this is stored in the database.
+This means over time, we get a better picture of which documents do or do not answer specific queries.
+
+When a new user question comes in, we can compare this question to previous questions, and exclude any suggested documents where the response was considered unhelpful.
+
+[source,cypher,rel=noplay]
+.Excluding Content
+----
+// Find 10 previous questions for similar questions (> 0.9)
+CALL db.index.vector.queryNodes('questions', 1000, $embedding)
+YIELD node AS unhelpful, score
+WHERE score >= 0.9 AND unhelpful:UnhelpfulResponse
+
+// Find suggested section where the response was marked as unhelpful
+MATCH (unhelpful)-[:SUGGESTED_SECTION]->(section)
+
+WITH doc, count(*) AS occurrences
```

Comment:
where is `doc` coming from?
Wouldn't it for this example be easier to just exclude (or re-order) the sections that had been unhelpful to responses in the past (ordered by unhelpfulness count).